Citing BleTIES and dependencies
===============================

BleTIES is research software. If you use it in a publication, please cite us in
the following way:

(Citation pending)

In addition, you should also cite the following dependencies that BleTIES is
built upon: 

 * `pysam` - A Heger, K Jacobs, et al. [https://github.com/pysam-developers/pysam ](https://github.com/pysam-developers/pysam)
 * `htslib`, `samtools` - H Li, et al. 2009. ["The Sequence Alignment/Map format and SAMtools"](https://doi.org/10.1093/bioinformatics/btp352) _Bioinformatics_ 25 (16) : 2078-2079.
 * `biopython` - PJA Cock, et al. 2009. ["Biopython: freely available Python tools for computational molecular biology and bioinformatics"](https://doi.org/10.1093/bioinformatics/btp163) _Bioinformatics_ 25 (11) : 1422-1423.
 * `muscle` - RC Edgar, 2004. ["MUSCLE: multiple sequence alignment with high accuracy and high throughput"](https://doi.org/10.1093/nar/gkh340) _Nucleic Acids Research_ 32 (5) : 1792-1797.
 * `ncrf` - RS Harris, M Cechova, KD Makova, 2019. ["Noise-cancelling repeat finder: uncovering tandem repeats in error-prone long-read sequencing data"](https://doi.org/10.1093/bioinformatics/btz484) _Bioinformatics_ 35 (22) : 4809-4811.
 * `spoa` - R Vaser, I Sovic, N Nagarajan, Mile Sikic, 2017. ["Fast and accurate de novo genome assembly from long uncorrected reads"](https://doi.org/10.1101/gr.214270.116) _Genome Research_ 27 : 737-746.


BibTeX
------

(Citation pending)
